### DISCUSSION TITLE
Would it be possible to animate IFC Objects ?

### DISCUSSION THREAD

**Discussion by PIUM-arc** [IfcOpenShell/IfcOpenShell#5559] [Q&A]
Hello,
I'm curious about the possibilities that IfcOpenShell could offer. Do you think it would be possible animate IFC objects of the model ? not moving them (at least not now) but change the color of an object in runtime, hide or show an object, add a text or a picture. 

I'm curious to have yours opinions on this
Thank you 
BR

**Comment by duarteframos** [comment]
Bonsai can do construction animations based on IFC scheduling data. There are various [examples scattered around the web ](https://community.osarch.org/discussion/1701/blenderbim-4d-crash-course. ). Other than that, I'm not aware of any further animation capabilities. 

Do you need it to specifically IFCOpenShell based? Why not just use directly Blender's animation capabilities through the Python API on the resulting Blender objects from the IFC entities and save as a .blend instead?

**Comment by Moult** [comment]
IFC as a schema does not support animation. So anything more complex than procedural animation based on time data in the model is not provided by Bonsai. That said, you can just animate things in Blender :)

**Comment by PIUM-arc** [comment]
Thank you for your responses.
Perhaps I should provide some context: imagine a supervision software where you retrieve data from the field, such as from PLCs, and use it to animate objects in your views. You have objects, shapes, symbols, and controls in your views that are animated based on this real-time data.

One common use case for supervision software is a Building Management System (BMS), where you can manage every aspect of your building directly from the system, such as electrical consumption, temperature, occupancy, and so on. People typically use 2D maps in their views, but one idea could be to use the IFC file of the building instead.

In this scenario, you would have a control in your view that acts as an IFC viewer, and you could animate IFC objects based on real-time data. For example, you might color an object red to indicate a fault, display a picture of someone in a specific room, or flash colors in the case of an alarm. These would be real-time animations without modifying the IFC file itself.

This could be done through an ActiveX control or a web component that can be opened in a web browser within the view.

I’m really curious to hear your thoughts on this idea and whether you see IfcOpenShell as a possible solution.

Thank you.

**Reply by Moult** [reply]
For your usecase, we do support visualising any property with coloured legends. However, as for live-updating (i.e. animated) in line with telemetry data, this is not possible because there is no standard for the storage of live timeseries data typically outside IFC.

**Reply by aothms** [reply]
Yeah 10y ago I hoped IFC-HDF5 could become this https://docs.hdfgroup.org/archive/support/HDF5/doc/HL/RM_H5PT.html where you have both performant append only timeseries storage, combined with the rich datatypes required for modelling the relational constructs in express. While still having something portable that you can handover. But I guess that ship has sailed, nobody wants this anymore based on HDF5.

But I think it would be good if we can as a community form a stance on how we envision people can deal with the combination of IFC and bulk data and slowly work towards some form of integrations.

**Comment by duarteframos** [comment]
The way I see it, IfcOpenShell is a set of libraries and tools to read, write, and process IFC files, it doesn't have any native visualization or animation facilities of its own.

What you seem to want to achieve is animate some sort of overlays or data over IFC objects based on external data. That would be, the way I see it up to whatever program you are using to visualize the IFC files, not the job of IFCOpenShell per se.
So the question is, what program are you intending to use to visualize those IFCs? Some sort of AR App, some web based viewer, a game engine, something else?

If Blender through the Bonsai addon, then yes, it sounds feasible to use Blender's Python API to animate objects, positions or colors based on external data.
